spec/lib/workflow/stores/mongo-spec.js

```diff
@@ -619,32 +599,262 @@ describe('Task Graph mongo store interface', function () {
+            dependencies:{
+                anyOf:{}
+            }
         };
 
-        var query = {
+        dependentTask.dependencies.anyOf[data.taskId] = 'failed';
+        dependentTask.dependencies.anyOf[uuid.v4()] = 'succeeded';
+
+        var queryGen = {
+            graphId: data.graphId,
+            reachable: true
+        };
+        queryGen['dependencies.' + data.taskId] = {
+            $in: _.difference(Constants.Task.FinishedStates, [data.state])
+        };
+        
+        var queryAny = {
             graphId: data.graphId,
             reachable: true
         };
-        query['dependencies.' + data.taskId] = {
+        queryAny['dependencies.anyOf.' + data.taskId] = {
             $in: _.difference(Constants.Task.FinishedStates, [data.state])
         };
+        
+        var dbFind = sinon.stub();
+        dbFind.withArgs(queryAny).resolves([
+            dependentTask
+        ]);
+        dbFind.withArgs(queryGen).resolves([]);
+
+        var waterlineUnderAnyOf = {
+            taskdependencies:{
+                find: dbFind,
+                //make sure the find cannot find anything under dependent.taskId
+                updateMongo: sinon.stub().resolves()
+            }
+        };
+
+        helper.setupInjector([
+            helper.require('/lib/workflow/stores/mongo'),
+            helper.di.simpleWrapper(waterlineUnderAnyOf, 'Services.Waterline')
+        ]);
+        mongo = helper.injector.get('TaskGraph.Stores.Mongo');
+        waterline = helper.injector.get('Services.Waterline');
+
+        var updateQuery = {
+            $or:[],
+            graphId: data.graphId,
+            reachable: true
+        };        
+        var queryEntry = {
+            taskId: dependentTask.taskId
+        };
+        updateQuery.$or.push(queryEntry);
+
+        var dbUpdate = {
+            $unset: {}
+        };
+        dbUpdate.$unset["dependencies.anyOf." + data.taskId] = "";
 
         return mongo.updateUnreachableTasks(data)
         .then(function() {
-            expect(waterline.taskdependencies.updateMongo).to.have.been.calledOnce;
+            expect(waterline.taskdependencies.find).to.have.been.called.Twice;
             expect(waterline.taskdependencies.updateMongo).to.have.been.calledWith(
-                query,
-                { $set: { reachable: false } },
-                { multi: true }
-            );
+                updateQuery,
+                dbUpdate,
+                {multi: true}
+            ); 
         });
     });
 
+    it('marks task unreachable when all tasks under anyOf faild', function(){
+        var data = {
+            state: 'succeeded',
+            taskId: uuid.v4(),
+            graphId: uuid.v4()
+        };       
+        
+        var dependentTask = {
+            taskId: uuid.v4(),
+            dependencies:{
+                anyOf:{}
+            }
+        };
+
+        dependentTask.dependencies.anyOf[data.taskId] = 'failed';
```

Comment:
Since the test condition is `when all tasks under anyOf faild`, so the `anyOf` should contain at least 2 tasks.